assets/src/hooks/useDetour.ts

```diff
@@ -26,35 +28,64 @@ export enum DetourState {
 }
 
 export const useDetour = ({ routePatternId, shape }: OriginalRoute) => {
-  const [state, setState] = useState<DetourState>(DetourState.Edit)
+  const [snapshot, send] = useMachine(createDetourMachine)
+
+  /*
+   * There's probably a better way to do this? Tags or maybe context?
+   * Tags seem more appropriate, but weird to manage Out-Of-Bounds state via tags.
+   * Maybe this entire API call could be moved to and managed by an actor
+   * combined with parallel child states within the state machine?
+   * -- https://stately.ai/docs/promise-actors */
+  const isFinishedDrawing = snapshot.matches({
+    "Detour Drawing": { Editing: "Finished Drawing" },
+  })
+  const isSharingDetour = snapshot.matches({ "Detour Drawing": "Share Detour" })
+  const isInFinishedDetourState = isFinishedDrawing || isSharingDetour
 
-  const [startPoint, setStartPoint] = useState<ShapePoint | null>(null)
-  const [endPoint, setEndPoint] = useState<ShapePoint | null>(null)
-  const [waypoints, setWaypoints] = useState<ShapePoint[]>([])
+  const firstWaypoint = snapshot.context.waypoints.at(0)
+  const lastWaypoint = snapshot.context.waypoints.at(-1)
+  // Lets also just assert that we're not operating on the same array element
+  const has2Waypoints = snapshot.context.waypoints.length >= 2
 
   const { result: finishedDetour } = useApiCall({
     apiCall: useCallback(async () => {
-      if (startPoint && endPoint) {
-        return fetchFinishedDetour(routePatternId, startPoint, endPoint)
-      } else {
+      if (!isInFinishedDetourState) {
         return null
       }
-    }, [startPoint, endPoint, routePatternId]),
+
+      /* Until we have "typegen" in XState,
+       * we need to validate these exist for typescript
+       *
+       * > [Warning] XState Typegen does not fully support XState v5 yet. However,
+       * > strongly-typed machines can still be achieved without Typegen.
+       * > -- https://stately.ai/docs/migration#use-typestypegen-instead-of-tstypes
+       */
+      if (
```

Comment:
Near the bottom, we're still computing `startPoint` and `endPoint` (using a slightly different formula that I _think_ will always give the same answer, assuming the state machine states are set up correctly).

Why not just compute `startPoint` and `endPoint` as values computed from `snapshot`? So something like:


```tsx
const startPoint = snapshot.context.waypoints.at(0)
const endPoint = snapshot.matches({
                   "Detour Drawing": { Editing: "Finished Drawing" },
                 }) && snapshot.context.waypoints.at(-1)

/* ... */

if (startPoint && endPoint) {
  return fetchFinishedDetour(routePatternId, startPoint, endPoint)
}

/* ... */

return {
  startPoint,

  endPoint: endPoint || undefined,
 
  /* ... */
}
```

Reply:
I tried to do this in [`9c4e8a9` (#2607)](https://github.com/mbta/skate/pull/2607/commits/9c4e8a9ba94c1ab0c1861c4659740905903664b7)
Let me know what you think!